service/src/main/java/bio/terra/cbas/dependencies/wes/CromwellService.java

```diff
@@ -83,18 +83,27 @@ public WorkflowQueryResult runSummary(String runId) throws ApiException {
 
   public Object getOutputs(String id)
       throws ApiException, DependencyNotAvailableException, AzureAccessTokenException {
-    return cromwellClient.wesAPI().getRunLog(id).getOutputs();
+    ApiClient client = cromwellClient.getWriteApiClient();
```

Comment:
(just to make sure we're on the same page: I'm operating under the assumption that these functions should use the "write client" when an operation is only available to creators/writers, and should use the "read client" when the operation is available to all user types)

should outputs be writer/creator only?

Reply:
Well, just thinking about this function in a vacuum: it seems to be reading some run information, which feels like it should be doable by readers. 

How `updateOutputAttributes`  uses this information feels like it should be a separate matter. Naively, I'd think that `updateOutputAttributes` should have its own way of enforcing "writers only"

Reply:
I see that `updateOutputAttributes` is being used by the `SmartRunsPoller`, and that `SmartRunsPoller` is the object that owns the Cromwell client. Thinking aloud: if a user is a reader, and they're on the run details page (or whichever page shows run status), is their browser being updated by a `SmartRunsPoller` with a reader client or a writer client?

Reply:
I hope I'm not leading you down the wrong path, but I'd be curious about how this would work... I'll ask at tech talk tomorrow, unless @cjllanwarne or @salonishah11 see this first

Reply:
My assumption is that a reader can see what results the poller is reporting, but the work being done by the poller itself is being done on behalf of a cromwell belonging to a user that had permissions to do the actual functions. So the `SmartRunsPoller` doesn't directly update the browser, `SmartRunsPoller` updates the information in the databases that is then pulled down in the UI which continues to refresh the information. That's my understanding, but maybe it's not totally correct.

Reply:
> Well, just thinking about this function in a vacuum: it seems to be reading some run information, which feels like it should be doable by readers.
> 
> How `updateOutputAttributes` uses this information feels like it should be a separate matter. Naively, I'd think that `updateOutputAttributes` should have its own way of enforcing "writers only"

I see what you mean, I think my logic here was that we are currently only using this api call in a place that writes back to the database, which is an action only writers should be able to do. For that reason, when we get the api client that will call the `getOutputs()` endpoint, we want that action to be taken by someone with write permissions. We may be able to enforce Sam permissions here, but I'd be curious as to what other people's perspectives are.

Reply:
The SmartRunsPoller is also being modified by the workflow completion callback API, not sure if that has an impact on the relevance of this discussion

Reply:
My $0.02: I would lean towards this being a "read client" action.

1. If it's being done during output updating in WDS it doesn't matter which Cromwell we read from
2. Reading from the "reader" should be significantly faster than the runner (no Leo/Sam/token/cross-network hops/.../etc to worry about)
3. If we reuse this in the future to allow a passthrough "getOutputs" endpoint in CBAS accessible to readers, it makes more sense to go to the reader Cromwell in the WORKFLOWS app.
4. There's a chance that we start shutting down the cromwell_runner app when the workflow is completed, but the reader Cromwell in the WORKFLOWS app should always be there to read outputs from

Reply:
I agree with what Chris said. I think the way I am thinking this that is if we need to look up any metadata or outputs or do any read from database related to a workflow, we want to use the Cromwell reader / read client so that it is accessible to the users without auth. The things we would go to user's Cromwell Runner is when submitting a workflow or aborting a workflow because user's individual Cromwell would be performing that action.